caffe2/operators/layer_norm_op.cc

```diff
@@ -186,16 +186,21 @@ to the end.)
 
 // Register layer norm with c10
 namespace {
+struct State final : public c10::KernelState {
+    at::optional<at::Tensor> scale = at::nullopt;
+    at::optional<at::Tensor> bias = at::nullopt;
+};
+
 template <class DataType>
-c10::IValue layer_norm_c10(c10::ArrayRef<c10::IValue> inputs) {
+c10::IValue layer_norm_c10(c10::ArrayRef<c10::IValue> inputs, c10::KernelState* state) { // TODO Pass in correct state type
   caffe2::Tensor X{c10::C10Tensor(inputs[0].toTensor())};
   caffe2::Tensor Y{c10::C10Tensor(inputs[1].toTensor())};
   caffe2::Tensor mean{c10::C10Tensor(inputs[2].toTensor())};
   caffe2::Tensor sig{c10::C10Tensor(inputs[3].toTensor())};
   int64_t axis = inputs[4].toInt();
   float epsilon = inputs[5].toDouble();
   caffe2::CPUContext context;
-  c10::core::opschema::LayerNorm::Cache* cache = inputs[6].toBlob()->GetMutable<c10::core::opschema::LayerNorm::Cache>();
+  State* cache = static_cast<State*>(state);
```

Comment:
Why can't the kernel be called with the correct state type in the first place? It's known to the registerer?

Reply:
As it is right now, this function can be stored directly into the dispatcher hashtable. If it was the correct state type, it'd need a wrapper function. It's probably fine since the wrapper function should (hopefully) inline the actual kernel function and avoid the indirection, but I wanted to test this first before implementing it. That's what the TODO above is for.